### ISSUE TITLE
NodePortLocal - Pools populated incorrectly when using ConfigMap

### ISSUE THREAD

**Issue by sectoreleven** [F5Networks/k8s-bigip-ctlr#2326]
#### Setup Details
 CIS Version : 2.8.0      
 Build: f5networks/k8s-bigip-ctlr:latest       
 BIGIP Version: Big IP 13.1.3.5      
 AS3 Version: 3.33      
 Agent Mode: AS3       
 Orchestration: K8S      
 Orchestration Version: Tanzu 1.21.8, Antrea 1.4       
 Pool Mode: NodePortLocal
 Additional Setup details: VMware Tanzu/Antrea

#### Description  
When using the new Tanzu/NodePortLocal implementation in v2.8.0 in conjunction with ConfigMap, pools are incorrectly populated with pod IPs and ports instead of Node IPs and local node ports.  Pools do get the correct number of members (one per pod/replica.)

#### Steps To Reproduce
1) Deploy the latest CIS from f5networks/k8s-bigip-ctlr:latest
2) Deploy a sample application with multiple replicas (such as f5devcentral/f5-hello-world:latest)
3) Deploy a ConfigMap containing an AS3 declaration.
4) Deploy a ClusterIP service annotated for CIS ConfigMap and NodePortLocal:

#### Expected Result 
The BIG-IP pools get created according to the ConfigMap definition, and populated one per pod, using node IPs and local node ports.

#### Actual Result  
The BIG-IP pools are created according to the ConfigMap definition, and populated one per pod but using pod private IPs and pod ports.

#### Diagnostic Information
[01-cis-deploy.yaml.txt](https://github.com/F5Networks/k8s-bigip-ctlr/files/8365774/01-cis-deploy.yaml.txt)
[02-deployment-hello-world.yaml.txt](https://github.com/F5Networks/k8s-bigip-ctlr/files/8365775/02-deployment-hello-world.yaml.txt)
[03-configmap-hello-world.yaml.txt](https://github.com/F5Networks/k8s-bigip-ctlr/files/8365776/03-configmap-hello-world.yaml.txt)
[04-clusterip-service-hello-world.yaml.txt](https://github.com/F5Networks/k8s-bigip-ctlr/files/8365777/04-clusterip-service-hello-world.yaml.txt)

Pod Details:
![image](https://user-images.githubusercontent.com/5799014/160474912-319109a6-2582-4d6d-99a8-fe10ab229eb4.png)

Pool Details:
![image](https://user-images.githubusercontent.com/5799014/160475048-4d62aa1c-7e6b-4c35-97fb-2f9826a9d311.png)

**Comment by mdditt2000** [comment]
ConfigMap doesn't work with NodePortLocal. Please use CRDs or Ingress. If you cannot use CRDs or Ingress, please let me know.

**Comment by sectoreleven** [comment]
No, we are unable to use CRDs nor Ingress because neither of them yet provide us the level of fine-grained control upon which we currently rely at the BIG-IP level - however, AS3 and ConfigMap do. 

For example, in conjunction with process/pod-level load-balancing using NodePortLocal, we also need to be able to specify the load-balancing method on the pool. Our company standard is _least-connections-member_ so changing to _round-robin_ would represent a fairly significant traffic and performance shift. I've found and added my voice in support of https://github.com/F5Networks/k8s-bigip-ctlr/issues/2304 for adding the LB-method feature to TransportServer and VirtualServer CRD.

In short - for applications that we're migrating from bare-metal/VM platforms to Kubernetes (and also in conjunction with an AS3 pipeline we have for configs _outside_ of Kubernetes), this level of control is necessary to ensure application performance and behavior are unchanged - so ConfigMap is our path forward.

**Comment by trinaths** [comment]
Created [CONTCNTR-3252] for internal tracking.

**Comment by sectoreleven** [comment]
Some of the additional functionality we need to achieve parity with our existing solution would be:

- SNAT Pool configuration for the BIG-IP virtual server
- OneConnect
- Throughput limiting (Bandwidth Controllers)
- Connection Rate limiting

The first two are more important than the latter.

In speaking with our SE, we might be able to work around the latter two by attaching an iRule to the virtual server using Policy CRD, then applying the limiting behavior at runtime, but that feels overly complex.

**Comment by mdditt2000** [comment]
SNAT Pool configuration for the BIG-IP virtual server - 

Current VirtualServer/TransportServer CRD can define snat pool existed in BIG-IP. 

```
snat: "/Common/testsnat"
```
---
OneConnect - CONTCNTR-2542 - Planned for CIS 2.9

Can you please send me the configmap or ltm configuration for 

Throughput limiting (Bandwidth Controllers)
Connection Rate limiting

**Comment by sectoreleven** [comment]
For connection limiting, we capture the data required to configure the following properties on the virtual server (https://clouddocs.f5.com/cli/tmsh-reference/v15/modules/ltm/ltm_virtual.html):

`connection-limit` - non-negative integer
`rate-limit` - non-negative integer
`rate-limit-mode` - enum (we only use `object` and `object-source` from the possible values on the platform)
`rate-limit-src-mask` - non-negative intger

Our current code is in iApps, so we have an `iapp::conf` call in the Tcl that looks like:
```
iapp::conf create ltm virtual ${virtual_name} \
  destination ${virtual_address}:${virtual_port} \
  connection-limit ${conn-limit-value} \
  rate-limit ${conn-ratelimit-value} \
  rate-limit-src-mask ${ratelimit-src-mask} \
  rate-limit-mode ${ratelimit_mode} \
```

For throughput limiting, our current use case focuses just on on static bandwidth controllers for device-level management - we're not doing any dynamic control yet, as our goal is to just prevent noisy-neighbor problems when a given virtual server on the LTM runs hot. For that, we collect:

`max-rate` - bandwidth value
`ip-tos` - defaults to `pass-through`
`measure` - we default to `enabled`
`link-qos` - defaults to `pass-through`
`max-user-rate` - bandwidth value
`max-user-rate-pps` - defaults to `0` (not configured)

And the corresponding `iapp::conf` with the remaing values supplied by constants/generated values looks like:
```
iapp::conf create net bwc policy /Common/${net_static_bwc_policy(bwc_name)} \
        description ${net_static_bwc_policy(description)} \
        max-rate ${net_static_bwc_policy(maximum_bwc_rate)} \
        dynamic ${net_static_bwc_policy(dynamic_bwc_flag)} \
        ip-tos ${net_static_bwc_policy(marking_tos)} \
        measure ${net_static_bwc_policy(measure_flag)} \
        link-qos ${net_static_bwc_policy(marking_link_qos)} \
        max-user-rate ${net_static_bwc_policy(maximum_rate_per_user)} \
        max-user-rate-pps ${net_static_bwc_policy(maximum_user_rate_pps)}
```

**Comment by sectoreleven** [comment]
And just to bring this back to the forefront - while all the other CRD functionality we've discussed is desired, beyond anything else we want the NodePortLocal functionality to work with ConfigMap, first.

> For example, in conjunction with process/pod-level load-balancing using NodePortLocal, we also need to be able to specify the load-balancing method on the pool. Our company standard is _least-connections-member_ so changing to _round-robin_ would represent a fairly significant traffic and performance shift. 

> In short - for applications that we're migrating from bare-metal/VM platforms to Kubernetes (and also in conjunction with an AS3 pipeline we have for configs outside of Kubernetes), this level of control is necessary to ensure application performance and behavior are unchanged - so ConfigMap is our path forward.
